File: packages/core/src/core/exceptions.ts
/**
 * Hierarchical exception classes for the MSF framework
 * @module core/exceptions
 */

/**
 * Severity levels for exceptions
 */
export enum SeverityLevel {
  /** Low severity, informational */
  LOW = 'low',
  /** Medium severity, requires attention */
  MEDIUM = 'medium',
  /** High severity, requires immediate action */
  HIGH = 'high',
  /** Critical severity, system-threatening */
  CRITICAL = 'critical',
}

/**
 * Base exception dictionary interface
 */
export interface ExceptionDict {
  /** Exception name */
  name: string;
  /** Error message */
  message: string;
  /** Error code */
  code: string;
  /** Severity level */
  severity: SeverityLevel;
  /** Stack trace */
  stack?: string;
  /** Additional context */
  context?: Record<string, unknown>;
  /** Timestamp */
  timestamp: string;
  /** Cause of the error */
  cause?: string;
}

/**
 * Base MSF error class with structured error information
 */
export class MSFError extends Error {
  /** Unique error code */
  public readonly code: string;
  /** Severity level */
  public readonly severity: SeverityLevel;
  /** Additional context data */
  public readonly context: Record<string, unknown>;
  /** Timestamp when error was created */
  public readonly timestamp: Date;

  /**
   * Create a new MSFError
   * @param message - Error message
   * @param code - Error code
   * @param severity - Severity level
   * @param context - Additional context
   * @param options - Error options including cause
   */
  constructor(
    message: string,
    code: string = 'MSF_ERROR',
    severity: SeverityLevel = SeverityLevel.MEDIUM,
    context: Record<string, unknown> = {},
    options?: { cause?: Error }
  ) {
    super(message, options);
    this.name = 'MSFError';
    this.code = code;
    this.severity = severity;
    this.context = context;
    this.timestamp = new Date();
  }

  /**
   * Convert error to dictionary representation
   * @returns Serializable error dictionary
   * @example
   * ```typescript
   * const dict = error.toDict();
   * console.log(dict.code, dict.message);
   * ```
   */
  toDict(): ExceptionDict {
    return {
      name: this.name,
      message: this.message,
      code: this.code,
      severity: this.severity,
      stack: this.stack,
      context: this.context,
      timestamp: this.timestamp.toISOString(),
      cause: this.cause?.toString(),
    };
  }
}

/**
 * Security-related error
 */
export class SecurityError extends MSFError {
  /**
   * Create a new SecurityError
   * @param message - Error message
   * @param context - Additional context
   * @param options - Error options
   */
  constructor(message: string, context: Record<string, unknown> = {}, options?: { cause?: Error }) {
    super(message, 'SECURITY_ERROR', SeverityLevel.HIGH, context, options);
    this.name = 'SecurityError';
  }
}

/**
 * Authentication failure error
 */
export class AuthenticationError extends MSFError {
  /**
   * Create a new AuthenticationError
   * @param message - Error message
   * @param context - Additional context
   * @param options - Error options
   */
  constructor(message: string, context: Record<string, unknown> = {}, options?: { cause?: Error }) {
    super(message, 'AUTHENTICATION_ERROR', SeverityLevel.HIGH, context, options);
    this.name = 'AuthenticationError';
  }
}

/**
 * Authorization failure error
 */
export class AuthorizationError extends MSFError {
  /**
   * Create a new AuthorizationError
   * @param message - Error message
   * @param context - Additional context
   * @param options - Error options
   */
  constructor(message: string, context: Record<string, unknown> = {}, options?: { cause?: Error }) {
    super(message, 'AUTHORIZATION_ERROR', SeverityLevel.HIGH, context, options);
    this.name = 'AuthorizationError';
  }
}

/**
 * Validation failure error
 */
export class ValidationError extends MSFError {
  /**
   * Create a new ValidationError
   * @param message - Error message
   * @param context - Additional context
   * @param options - Error options
   */
  constructor(message: string, context: Record<string, unknown> = {}, options?: { cause?: Error }) {
    super(message, 'VALIDATION_ERROR', SeverityLevel.MEDIUM, context, options);
    this.name = 'ValidationError';
  }
}

/**
 * Cryptography operation error
 */
export class CryptographyError extends MSFError {
  /**
   * Create a new CryptographyError
   * @param message - Error message
   * @param context - Additional context
   * @param options - Error options
   */
  constructor(message: string, context: Record<string, unknown> = {}, options?: { cause?: Error }) {
    super(message, 'CRYPTOGRAPHY_ERROR', SeverityLevel.CRITICAL, context, options);
    this.name = 'CryptographyError';
  }
}

/**
 * Rate limit exceeded error
 */
export class RateLimitError extends MSFError {
  /** Retry-after duration in seconds */
  public readonly retryAfter: number;

  /**
   * Create a new RateLimitError
   * @param message - Error message
   * @param retryAfter - Seconds until retry is allowed
   * @param context - Additional context
   */
  constructor(message: string, retryAfter: number = 60, context: Record<string, unknown> = {}) {
    super(message, 'RATE_LIMIT_ERROR', SeverityLevel.MEDIUM, { retryAfter, ...context });
    this.name = 'RateLimitError';
    this.retryAfter = retryAfter;
  }
}

/**
 * Operation timeout error
 */
export class TimeoutError extends MSFError {
  /** Timeout duration in milliseconds */
  public readonly timeoutMs: number;

  /**
   * Create a new TimeoutError
   * @param message - Error message
   * @param timeoutMs - Timeout duration
   * @param context - Additional context
   */
  constructor(message: string, timeoutMs: number, context: Record<string, unknown> = {}) {
    super(message, 'TIMEOUT_ERROR', SeverityLevel.HIGH, { timeoutMs, ...context });
    this.name = 'TimeoutError';
    this.timeoutMs = timeoutMs;
  }
}

/**
 * Policy violation error
 */
export class PolicyViolationError extends MSFError {
  /** Policy rule that was violated */
  public readonly policyRule: string;

  /**
   * Create a new PolicyViolationError
   * @param message - Error message
   * @param policyRule - The violated policy rule
   * @param context - Additional context
   */
  constructor(message: string, policyRule: string, context: Record<string, unknown> = {}) {
    super(message, 'POLICY_VIOLATION', SeverityLevel.HIGH, { policyRule, ...context });
    this.name = 'PolicyViolationError';
    this.policyRule = policyRule;
  }
}

/**
 * Threat detection error
 */
export class ThreatDetectedError extends MSFError {
  /** Threat type identifier */
  public readonly threatType: string;
  /** MITRE ATT&CK technique ID if applicable */
  public readonly mitreTechnique?: string;

  /**
   * Create a new ThreatDetectedError
   * @param message - Error message
   * @param threatType - Type of threat detected
   * @param mitreTechnique - MITRE ATT&CK technique ID
   * @param context - Additional context
   */
  constructor(
    message: string,
    threatType: string,
    mitreTechnique?: string,
    context: Record<string, unknown> = {}
  ) {
    super(message, 'THREAT_DETECTED', SeverityLevel.CRITICAL, { threatType, mitreTechnique, ...context });
    this.name = 'ThreatDetectedError';
    this.threatType = threatType;
    this.mitreTechnique = mitreTechnique;
  }
}

/**
 * Configuration error
 */
export class ConfigurationError extends MSFError {
  /**
   * Create a new ConfigurationError
   * @param message - Error message
   * @param context - Additional context
   * @param options - Error options
   */
  constructor(message: string, context: Record<string, unknown> = {}, options?: { cause?: Error }) {
    super(message, 'CONFIGURATION_ERROR', SeverityLevel.HIGH, context, options);
    this.name = 'ConfigurationError';
  }
}

/**
 * Plugin-related error
 */
export class PluginError extends MSFError {
  /** Plugin name */
  public readonly pluginName: string;

  /**
   * Create a new PluginError
   * @param message - Error message
   * @param pluginName - Name of the plugin
   * @param context - Additional context
   * @param options - Error options
   */
  constructor(
    message: string,
    pluginName: string,
    context: Record<string, unknown> = {},
    options?: { cause?: Error }
  ) {
    super(message, 'PLUGIN_ERROR', SeverityLevel.HIGH, { pluginName, ...context }, options);
    this.name = 'PluginError';
    this.pluginName = pluginName;
  }
}

File: master_security_python/src/master_security/core/exceptions.py
"""
Master Security Framework - Exception Hierarchy
================================================

Typed exception hierarchy for all security operations.
Each exception includes security context for tracing and alerting.

Exception Hierarchy:
    MSFError
    +-- SecurityError
    |   +-- AuthenticationError
    |   +-- AuthorizationError
    |   +-- ValidationError
    |   +-- CryptographyError
    |   +-- RateLimitError
    |   +-- TimeoutError
    |   +-- PolicyViolationError
    |   +-- ThreatDetectedError
    +-- ConfigurationError
    +-- PluginError

Usage:
    >>> from master_security.core import AuthenticationError
    >>> raise AuthenticationError("Invalid JWT", user_id="u123", severity="high")
"""

from __future__ import annotations

from typing import Any, Optional
from enum import Enum


class SeverityLevel(Enum):
    """Security event severity levels."""
    INFO = "info"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class MSFError(Exception):
    """
    Base exception for all Master Security Framework errors.

    Attributes:
        message: Human-readable error message
        code: Machine-readable error code
        severity: Security severity level
        context: Additional context dictionary
        trace_id: Unique trace identifier
    """

    def __init__(
        self,
        message: str,
        code: str = "MSF_ERROR",
        severity: SeverityLevel = SeverityLevel.MEDIUM,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
    ) -> None:
        super().__init__(message)
        self.message = message
        self.code = code
        self.severity = severity
        self.context = context or {}
        self.trace_id = trace_id or ""
        self.context["error_code"] = code
        self.context["severity"] = severity.value

    def to_dict(self) -> dict[str, Any]:
        """Convert exception to dictionary for logging."""
        return {
            "error": self.__class__.__name__,
            "message": self.message,
            "code": self.code,
            "severity": self.severity.value,
            "context": self.context,
            "trace_id": self.trace_id,
        }


class SecurityError(MSFError):
    """Base exception for security-related errors."""

    def __init__(
        self,
        message: str,
        code: str = "SECURITY_ERROR",
        severity: SeverityLevel = SeverityLevel.HIGH,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)


class AuthenticationError(SecurityError):
    """Authentication failure - invalid credentials, expired tokens, etc."""

    def __init__(
        self,
        message: str,
        code: str = "AUTH_ERROR",
        severity: SeverityLevel = SeverityLevel.HIGH,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)


class AuthorizationError(SecurityError):
    """Authorization failure - insufficient permissions, access denied."""

    def __init__(
        self,
        message: str,
        code: str = "AUTHZ_ERROR",
        severity: SeverityLevel = SeverityLevel.HIGH,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)


class ValidationError(SecurityError):
    """Input validation failure - malformed data, schema violations."""

    def __init__(
        self,
        message: str,
        code: str = "VALIDATION_ERROR",
        severity: SeverityLevel = SeverityLevel.MEDIUM,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)


class CryptographyError(SecurityError):
    """Cryptographic operation failure - encryption, decryption, signing."""

    def __init__(
        self,
        message: str,
        code: str = "CRYPTO_ERROR",
        severity: SeverityLevel = SeverityLevel.CRITICAL,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)


class RateLimitError(SecurityError):
    """Rate limit exceeded."""

    def __init__(
        self,
        message: str = "Rate limit exceeded",
        code: str = "RATE_LIMIT",
        severity: SeverityLevel = SeverityLevel.MEDIUM,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
        retry_after: int = 60,
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)
        self.retry_after = retry_after
        self.context["retry_after"] = retry_after


class TimeoutError(SecurityError):
    """Operation timeout - security operation took too long."""

    def __init__(
        self,
        message: str = "Security operation timeout",
        code: str = "TIMEOUT",
        severity: SeverityLevel = SeverityLevel.MEDIUM,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
        timeout_ms: int = 5000,
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)
        self.timeout_ms = timeout_ms
        self.context["timeout_ms"] = timeout_ms


class PolicyViolationError(SecurityError):
    """Security policy violation detected."""

    def __init__(
        self,
        message: str,
        code: str = "POLICY_VIOLATION",
        severity: SeverityLevel = SeverityLevel.HIGH,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
        policy_name: str = "",
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)
        self.policy_name = policy_name
        self.context["policy_name"] = policy_name


class ThreatDetectedError(SecurityError):
    """Active threat detected - requires immediate action."""

    def __init__(
        self,
        message: str,
        code: str = "THREAT_DETECTED",
        severity: SeverityLevel = SeverityLevel.CRITICAL,
        context: Optional[dict[str, Any]] = None,
        trace_id: Optional[str] = None,
        threat_type: str = "unknown",
        mitre_technique: str = "",
    ) -> None:
        super().__init__(message, code, severity, context, trace_id)
        self.threat_type = threat_type
        self.mitre_technique = mitre_technique
        self.context["threat_type"] = threat_type
        self.context["mitre_technique"] = mitre_technique


class ConfigurationError(MSFError):
    """Configuration error - invalid or missing configuration."""

    def __init__(
        self,
        message: str,
        code: str = "CONFIG_ERROR",
        context: Optional[dict[str, Any]] = None,
    ) -> None:
        super().__init__(message, code, SeverityLevel.HIGH, context)


class PluginError(MSFError):
    """Plugin error - plugin load, execution, or compatibility failure."""

    def __init__(
        self,
        message: str,
        code: str = "PLUGIN_ERROR",
        context: Optional[dict[str, Any]] = None,
        plugin_name: str = "",
    ) -> None:
        super().__init__(message, code, SeverityLevel.HIGH, context)
        self.plugin_name = plugin_name
        self.context["plugin_name"] = plugin_name
